File: exam_LRX/py/work12.py
import cv2
import numpy as np
import matplotlib.pyplot as plt
from matplotlib import rcParams

# 设置中文字体
rcParams['font.family'] = 'SimHei'  # 设置为黑体
rcParams['axes.unicode_minus'] = False  # 解决负号显示问题

# 读取输入图像
input_path = "../image/image10.png"  # 替换为您的文件路径
image = cv2.imread(input_path, cv2.IMREAD_GRAYSCALE)

# 二值化处理（确保图像为黑白）
_, binary_image = cv2.threshold(image, 128, 255, cv2.THRESH_BINARY)

# 定义结构元素（可根据需求调整大小和形状）
kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (10, 10))

# 应用闭运算（填充空洞）
closed_image = cv2.morphologyEx(binary_image, cv2.MORPH_CLOSE, kernel)

# 显示结果
plt.figure(figsize=(10, 5))

# 左图：原始图像
plt.subplot(1, 2, 1)
plt.imshow(binary_image, cmap="gray")
plt.title("原 图")
plt.axis("off")

# 右图：填充空洞后的图像
plt.subplot(1, 2, 2)
plt.imshow(closed_image, cmap="gray")
plt.title("填充孔洞后的图像")
plt.axis("off")

plt.tight_layout()
plt.show()

File: exam_LRX/cpp/work12.cpp
//
// Created by liryi on 24-11-24.
//
#include <opencv2/opencv.hpp>
#include <iostream>

using namespace cv;
using namespace std;

// 二值化图像
Mat binarizeImage(const Mat& image, double thresholdValue) {
    Mat binaryImage;
    threshold(image, binaryImage, thresholdValue, 255, THRESH_BINARY);
    return binaryImage;
}

// 应用闭运算（填充空洞）
Mat applyMorphClose(const Mat& binaryImage, const Size& kernelSize) {
    Mat kernel = getStructuringElement(MORPH_RECT, kernelSize);
    Mat closedImage;
    morphologyEx(binaryImage, closedImage, MORPH_CLOSE, kernel);
    return closedImage;
}

// 拼接并显示结果
void displayResults(const Mat& originalImage, const Mat& processedImage) {
    Mat display;
    hconcat(originalImage, processedImage, display);

    putText(display, "Original", Point(10, 30), FONT_HERSHEY_SIMPLEX, 1, Scalar(255), 2);
    putText(display, "Hole Filled", Point(originalImage.cols + 10, 30), FONT_HERSHEY_SIMPLEX, 1, Scalar(255), 2);

    namedWindow("Results", WINDOW_NORMAL);
    imshow("Results", display);
    waitKey(0);
}

int main() {
    string inputPath = "E:\\DL_class\\exam_LRX\\cpp\\images\\image10.png"; // 替换为您的图像路径
    Mat image = imread(inputPath, IMREAD_GRAYSCALE);

    if (image.empty()) {
        cerr << "Error: Unable to load image!" << endl;
        return -1;
    }

    // 二值化
    Mat binaryImage = binarizeImage(image, 128.0);

    // 闭运算
    Mat closedImage = applyMorphClose(binaryImage, Size(10, 10));

    // 显示结果
    displayResults(binaryImage, closedImage);

    return 0;
}

